### PR TITLE
teleport systemd unit upgrader

### PR DESCRIPTION
**Pull request by fspmarshall**
WIP upgrader. Currently does not actually perform upgrades (just being used to test state-machine and schedule format).

Related:
- https://github.com/gravitational/teleport/pull/23062
- https://github.com/gravitational/teleport/pull/22850

### CODE REVIEW

File: assets/systemd-unit-upgrader/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    utc_start_hour: 2
+    weekdays:
+      - Mon
+      - Wed
+      - Fri
+EOF
+$ tctl create -f mw.yaml
+```
+
+Second, a TLS endpoing must be set up to serve the "target version" for the updater (i.e.
```
Comment (webvictim) [inline assets/systemd-unit-upgrader/README.md:26]:
```suggestion
Second, a TLS endpoint must be set up to serve the "target version" for the updater (i.e.
```

File: assets/systemd-unit-upgrader/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+```bash
+$ teleport-upgrade 
+[!] fetching localhost:8000/version in insecure mode (not safe for production use). [ 186 ]
+[i] an upgrade is available (1.2.3 -> 1.2.4) [ 274 ]
+[!] fetching localhost:8000/critical in insecure mode (not safe for production use). [ 186 ]
+[i] agent may be unhealthy. marking for potential future upgrade. [ 290 ]
+```
+
+Note that log details will differ based on wether or not your agent has exported a maintenance window
```
Comment (webvictim) [inline assets/systemd-unit-upgrader/README.md:55]:
```suggestion
Note that log details will differ based on whether or not your agent has exported a maintenance window
```

File: assets/systemd-unit-upgrader/teleport-upgrade [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    log_debug "schedule appears unhealthy (contains no future windows)."
+    return 1
+}
+
+
+# get_current_version gets the current version of the teleport binary. if this
+# operation fails, we assume that teleport is eligible for upgrade.
+get_current_version() {
+    echo "1.2.3"
+    return 0
```
Comment (webvictim) [inline assets/systemd-unit-upgrader/teleport-upgrade:172]: I presume this is here for debugging?

File: assets/systemd-unit-upgrader/teleport-upgrade [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    else
+        log_info "an upgrade is available ($current_version -> $target_version)"
+    fi
+else
+    log_info "failed to detect current version, assuming $target_version to be valid target."
+fi
+
+
+if is_critical; then
+    log_info "upgrade is marked as critical, will be attempted without checking maintenenance window."
```
Comment (webvictim) [inline assets/systemd-unit-upgrader/teleport-upgrade:278]:
```suggestion
    log_info "upgrade is marked as critical, will be attempted without checking maintenance window."
```

File: assets/systemd-unit-upgrader/teleport-upgrade.service [hunk 5]
```diff
@@ -0,0 +1,5 @@
+[Unit]
+Description=sdt unit launched by timer
```
Comment (webvictim) [inline assets/systemd-unit-upgrader/teleport-upgrade.service:2]: What's `sdt` here?

**Review by webvictim — COMMENTED** [review]
The `bash` parts LGTM for a draft, assuming they work!

File: assets/systemd-unit-upgrader/teleport-upgrade [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+        echo "${tv_output#v}"
+        return 0
+    fi
+    return 1
+}
+
+# is_critical checks if the /critical endpoint is available and set to 'yes'.
+is_critical() {
+    if ef_output="$(upgrade_endpoint_fetch critical)"; then
+        if [[ $ef_output == "yes" ]]; then
```
Comment (fheinecke) [inline assets/systemd-unit-upgrader/teleport-upgrade:219]: You may want to trim whitespace (spaces, carriage returns, line feeds) and maybe do a case-insensitive comparison.

File: assets/systemd-unit-upgrader/teleport-upgrade [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        echo "$default_value"
+        return 0
+    fi
+
+    echo "$normalized_value"
+}
+
+
+# upgrade_endpoint is the base host+path used to load the /version and /critical endpoints
+upgrade_endpoint="$(get_conf endpoint update.gravitational.io/v1/stable/cloud)"
```
Comment (fheinecke) [inline assets/systemd-unit-upgrader/teleport-upgrade:35]:
```suggestion
upgrade_endpoint="$(get_conf endpoint updates.releases.teleport.dev/v1/stable/cloud)"
```

Note that I'm going to file a PR for a dev bucket at `updates.releases.development.teleport.dev` as well

**Comment by fspmarshall** [comment]
moved to https://github.com/gravitational/teleport.e/pull/1109

**Pull request closed without merging**